### PR TITLE
Update auto link logic

### PR DESCRIPTION
**Pull request by revam**

### CODE REVIEW

File: src/core/utilities/auto-match-regexes.ts [hunk 1]
```diff
@@ -4,7 +4,14 @@ const PathMatchRuleSet: PathMatchRule[] = [
...
+      /^(?:[{[(](?<releaseGroup>[a-zA-Z0-9_-]+)[)}\]][\s_]*)?(?<showName>.+?(?:[\s_]*\((?<year>(?:19|20)\d{2})\)[\s_]*)?)[\s_]+(?:-[\s_]+)?(?:(?:(?<isThemeSong>(?:NC)?(?:ED|OP)[\s_]*))?|(?<isSpecial>SP?)?|S(?<season>\d+)E|E?)(?<episode>\d+(?:-\d+|\.\d+)?)(?:[vV](?<version>\d{1,2}))?(?:[\s_]*-)?(?:[\s_]*(?:[([{](?:(?:(?:[uf]?hd|sd|bd|ld|dvd|blu-?ray)?(?:[48]k|\d{3,4}[pi])|(?:[uf]?hd|sd)|\d{3,4}x\d{3,4}|dual[\s_-]audio|(?:bd|dvd|ld|blu-?ray)(?:[\s_-]*rip)?|rip|(?:hevc|x26[45])|(?:ac3|aac|flac)(?:[\s_]*[257]\.[0124])?|(?:\w{2,3}[\s_-]*)?(?:sub|dub)(?:s|titles?)?|(?:un)?cen(?:\.|sored)?)[\s_]*)+[)\]}]|(?:(?:(?:[uf]?hd|sd|bd|ld|dvd|blu-?ray)?(?:[48]k|\d{3,4}[pi])|(?:[uf]?hd|sd)|\d{3,4}x\d{3,4}|dual[\s_-]audio|(?:bd|dvd|ld|blu-?ray)(?:[\s_-]*rip)?|rip|(?:hevc|x26[45])|(?:ac3|aac|flac)(?:[\s_]*[257]\.[0124])?|(?:\w{2,3}[\s_-]*)?(?:sub|dub)(?:s|titles?)?|(?:un)?cen(?:\.|sored)?)[\s_]*)+))*(?:[\s_]*\[(?<crc32>[0-9a-fA-F]{8})\])?[\s_]*\.(?<extension>[a-zA-Z0-9_\-+]+)$/i,
```
Comment (revam) [inline src/core/utilities/auto-match-regexes.ts:7]: I tested this while writing it.

File: src/pages/utilities/UnrecognizedUtilityTabs/LinkFilesTab.tsx [hunk 2]
```diff
@@ -252,7 +252,7 @@ function LinkFilesTab() {
   );
 
   const episodes = useMemo(() => anidbEpisodesQuery?.data?.List || [], [anidbEpisodesQuery]);
-  const fileLinks = useMemo(() =>
+  const orderedLinks = useMemo(() =>
```
Comment (revam) [inline src/pages/utilities/UnrecognizedUtilityTabs/LinkFilesTab.tsx:255]: I renamed it to make more sense for what it is. It's the links, but ordered. It can be reverted if any of you feel like it's an unnecessary change.

**Comment by revam** [comment]
I have tested the logic in limited capacity in the browser developer tools, and tested the regular expressions standalone by themselves.

File: src/core/utilities/auto-match-logic.ts [hunk 3]
```diff
@@ -75,28 +71,40 @@ export function detectShow(filePath: string | undefined | null): PathDetails | n
...
-        }
+      if (episodeEnd - episodeStart < 0) {
+        const tempEpisode = episodeStart;
+        episodeStart = episodeEnd;
+        episodeEnd = tempEpisode;
+      }
+
+      // Special handling of in-between episodes specials. We can't get the
+      // special episode number, but we can guess it based on context later
+      // provided the user tries to link all the episodes in the series at once.
```
Comment (revam) [inline src/core/utilities/auto-match-logic.ts:88]: missing a "when"

File: src/core/utilities/auto-match-logic.ts [hunk 4]
```diff
@@ -107,20 +115,24 @@ export function detectShow(filePath: string | undefined | null): PathDetails | n
...
         showName = parentMatch.groups.showName?.trim() || null;
-        if (showName && showName) show.showName = showName;
+        if (showName && showName) pathDetails.showName = showName;
       }
 
-      const finalShow = transform(show, match, parentMatch, grandParentMatch);
-      if (finalShow) {
-        return finalShow;
+      // Transform the details if the rule has a trasformer/validator.
+      pathDetails = transform(pathDetails, match, parentMatch, grandParentMatch);
```
Comment (hidden4003) [inline src/core/utilities/auto-match-logic.ts:130]: why does it have to overwrite itself?

File: src/core/utilities/auto-match-logic.ts [hunk 5]
```diff
@@ -132,7 +144,7 @@ export function findMostCommonShowName(showList: Array<PathDetails | null>): str
     return '';
   }
 
-  const showNameMap = reduce(showList, (acc, show) => {
+  const showNameMap = showList.reduce((acc, show) => {
```
Comment (hidden4003) [inline src/core/utilities/auto-match-logic.ts:147]: what is the point of replacing lodash ? It has to be consistent throughout the project.

Comment (revam) [reply src/core/utilities/auto-match-logic.ts:147]: I can add lodash back.

**Comment by revam** [comment]
`fileLinks` in the static render function needs to be updated to `orderedLinks` or the change tonrenane the variable needs to be rolled back.
